File: OrientationPFIEV/Forms/FrmAjouterCandidat.cs
using OrientationPFIEV.Data.Repositories;
using OrientationPFIEV.Data;
using OrientationPFIEV.Helpers;
using OrientationPFIEV.Models;

namespace OrientationPFIEV.Forms;

/// <summary>
/// Add / Edit dialog for a Candidat record.
/// Pass candidatId for edit mode; omit for add mode.
/// Suivant button saves and clears for the next candidate entry.
/// </summary>
public partial class FrmAjouterCandidat : Form
{
    private readonly CandidatRepository _repo;
    private readonly int? _existingId;
    private readonly EtablissementRepository _etabRepo;

    public FrmAjouterCandidat(CandidatRepository repo, int? candidatId = null)
    {
        InitializeComponent();
        _repo       = repo;
        _existingId = candidatId;
        _etabRepo   = new EtablissementRepository(ConfigDatabase.Context);
    }

    protected override void OnLoad(EventArgs e)
    {
        base.OnLoad(e);
        ApplyLocalization();
        LoadComboBoxes();
        if (_existingId.HasValue) PopulateFields(_existingId.Value);
    }

    // ── Localization ───────────────────────────────────────────────────────────

    private void ApplyLocalization()
    {
        Text             = L.Get("TitleAjouterCandidat");
        lblNom.Text      = L.Get("LblNom") + " :";
        lblNomInter.Text = "Nom intermédiaire :";
        lblPrenom.Text   = L.Get("ColPrenom") + " :";
        lblAnonymat.Text = L.Get("ColAnonymat") + " :";
        lblDOB.Text      = L.Get("LblDOB") + " :";
        lblSexe.Text     = L.Get("LblSex") + " :";
        lblStatut.Text   = L.Get("LblStatus") + " :";
        lblLangue.Text   = L.Get("LblLanguage") + " :";
        lblEtab.Text     = L.Get("LblEtablissement") + " :";
        btnOk.Text       = L.Get("BtnOk");
        btnAnnuler.Text  = L.Get("BtnCancel");
        btnSuivant.Text  = "Suivant";
        // Hide Suivant in edit mode — only useful for batch add
        btnSuivant.Visible = !_existingId.HasValue;
    }

    // ── Data ───────────────────────────────────────────────────────────────────

    private void LoadComboBoxes()
    {
        comboSexe.Items.Clear();
        comboSexe.Items.Add("M");
        comboSexe.Items.Add("F");
        comboSexe.SelectedIndex = 0;

        comboStatut.Items.Clear();
        comboStatut.Items.Add("I");
        comboStatut.Items.Add("E");
        comboStatut.SelectedIndex = 0;

        comboLangue.Items.Clear();
        comboLangue.Items.Add("fr");
        comboLangue.Items.Add("vi");
        comboLangue.SelectedIndex = 0;

        var etabs = _etabRepo.GetAll();
        comboEtab.DataSource    = etabs;
        comboEtab.DisplayMember = "EtabNom";
        comboEtab.ValueMember   = "EtabID";
        if (etabs.Count > 0) comboEtab.SelectedIndex = 0;
    }

    private void PopulateFields(int id)
    {
        var c = _repo.GetById(id);
        if (c == null) return;
        txtNom.Text              = c.Nom;
        txtNomInter.Text         = c.NomIntermediaire;
        txtPrenom.Text           = c.Prenom;
        txtAnonymat.Text         = c.Anonymat;
        if (c.DateDeNaissance.HasValue)
            dtpDateNaissance.Value = c.DateDeNaissance.Value;
        comboSexe.SelectedItem   = c.Sexe;
        comboStatut.SelectedItem = c.CandidatStatut;
        comboLangue.SelectedItem = c.Langue;
        // Match etab by ID
        foreach (Etablissement etab in comboEtab.Items)
        {
            if (etab.EtabID == c.EtabID)
            {
                comboEtab.SelectedItem = etab;
                break;
            }
        }
    }

    private bool ValidateInputs()
    {
        if (string.IsNullOrWhiteSpace(txtNom.Text) ||
            string.IsNullOrWhiteSpace(txtPrenom.Text))
        {
            MessageBox.Show(L.Get("MsgFieldRequired"), "",
                MessageBoxButtons.OK, MessageBoxIcon.Warning);
            return false;
        }
        return true;
    }

    private Candidat BuildCandidat() => new()
    {
        Nom              = txtNom.Text.Trim(),
        NomIntermediaire = txtNomInter.Text.Trim(),
        Prenom           = txtPrenom.Text.Trim(),
        Anonymat         = txtAnonymat.Text.Trim(),
        DateDeNaissance  = dtpDateNaissance.Value.Date,
        Sexe             = comboSexe.SelectedItem?.ToString() ?? "M",
        CandidatStatut   = comboStatut.SelectedItem?.ToString() ?? "I",
        Langue           = comboLangue.SelectedItem?.ToString() ?? "fr",
        EtabID           = comboEtab.SelectedValue is int eid ? eid : 0,
    };

    private void ClearForm()
    {
        txtNom.Text      = "";
        txtNomInter.Text = "";
        txtPrenom.Text   = "";
        txtAnonymat.Text = "";
        dtpDateNaissance.Value   = DateTime.Today;
        comboSexe.SelectedIndex  = 0;
        comboStatut.SelectedIndex = 0;
        comboLangue.SelectedIndex = 0;
        if (comboEtab.Items.Count > 0) comboEtab.SelectedIndex = 0;
        txtNom.Focus();
    }

    // ── Event handlers ─────────────────────────────────────────────────────────

    private void btnOk_Click(object sender, EventArgs e) => SaveAndClose(close: true);

    private void btnSuivant_Click(object sender, EventArgs e) => SaveAndClose(close: false);

    private void btnAnnuler_Click(object sender, EventArgs e)
    {
        DialogResult = DialogResult.Cancel;
        Close();
    }

    private void SaveAndClose(bool close)
    {
        if (!ValidateInputs()) return;
        var c = BuildCandidat();
        try
        {
            if (_existingId.HasValue)
            {
                c.CandidatID = _existingId.Value;
                _repo.Update(c);
            }
            else
            {
                _repo.Add(c);
            }

            if (close)
            {
                DialogResult = DialogResult.OK;
                Close();
            }
            else
            {
                ClearForm();
            }
        }
        catch (Exception ex)
        {
            MessageBox.Show(ex.Message, "", MessageBoxButtons.OK, MessageBoxIcon.Error);
        }
    }
}

File: app/views/frm_ajouter_candidat.py
from typing import Optional
from PySide6.QtWidgets import (
    QDialog, QFormLayout, QLineEdit, QComboBox, QDateEdit,
    QDialogButtonBox, QMessageBox
)
from PySide6.QtCore import QDate
from app.localization import L
from app.models.candidat import Candidat


class FrmAjouterCandidat(QDialog):
    def __init__(self, parent=None, candidat_id: Optional[int] = None):
        super().__init__(parent)
        self._candidat_id = candidat_id
        self._etabs = []
        title = L.get("TitleAjouterCandidat")
        self.setWindowTitle(title)
        self.setFixedSize(360, 320)
        self._setup_ui()
        self._load_etabs()
        if candidat_id:
            self._load_candidat(candidat_id)

    def _setup_ui(self):
        layout = QFormLayout(self)
        self._txt_nom = QLineEdit()
        self._txt_nom_int = QLineEdit()
        self._txt_prenom = QLineEdit()
        self._date_dob = QDateEdit()
        self._date_dob.setDisplayFormat("dd/MM/yyyy")
        self._date_dob.setCalendarPopup(True)
        self._date_dob.setDate(QDate.currentDate())
        self._combo_sexe = QComboBox()
        self._combo_sexe.addItems(["M", "F"])
        self._combo_statut = QComboBox()
        self._combo_statut.addItems(["I", "E"])
        self._combo_langue = QComboBox()
        self._combo_langue.addItems(["fr", "vi"])
        self._combo_etab = QComboBox()
        self._txt_anonymat = QLineEdit()

        layout.addRow(L.get("LblNom"), self._txt_nom)
        layout.addRow(L.get("LblNomIntermediaire"), self._txt_nom_int)
        layout.addRow(L.get("ColPrenom"), self._txt_prenom)
        layout.addRow(L.get("LblDOB"), self._date_dob)
        layout.addRow(L.get("LblSex"), self._combo_sexe)
        layout.addRow(L.get("LblStatus"), self._combo_statut)
        layout.addRow(L.get("LblLanguage"), self._combo_langue)
        layout.addRow(L.get("LblSchool"), self._combo_etab)
        layout.addRow(L.get("ColAnonymat"), self._txt_anonymat)

        btns = QDialogButtonBox(QDialogButtonBox.Ok | QDialogButtonBox.Cancel)
        btns.accepted.connect(self._on_ok)
        btns.rejected.connect(self.reject)
        layout.addRow(btns)

    def _load_etabs(self):
        from app.database.repositories.etablissement_repo import EtablissementRepository
        self._etabs = EtablissementRepository(use_session_db=True).get_all()
        self._combo_etab.clear()
        for e in self._etabs:
            self._combo_etab.addItem(e.etab_nom)

    def _load_candidat(self, candidat_id: int):
        from app.database.repositories.candidat_repo import CandidatRepository
        c = CandidatRepository().get_by_id(candidat_id)
        if not c:
            return
        self._txt_nom.setText(c.nom)
        self._txt_nom_int.setText(c.nom_intermediaire)
        self._txt_prenom.setText(c.prenom)
        if c.date_de_naissance:
            try:
                d = QDate.fromString(c.date_de_naissance, "yyyy-MM-dd")
                if d.isValid():
                    self._date_dob.setDate(d)
            except Exception:
                pass
        idx = self._combo_sexe.findText(c.sexe)
        if idx >= 0:
            self._combo_sexe.setCurrentIndex(idx)
        idx = self._combo_statut.findText(c.candidat_statut)
        if idx >= 0:
            self._combo_statut.setCurrentIndex(idx)
        idx = self._combo_langue.findText(c.langue)
        if idx >= 0:
            self._combo_langue.setCurrentIndex(idx)
        for i, e in enumerate(self._etabs):
            if e.etab_id == c.etab_id:
                self._combo_etab.setCurrentIndex(i)
                break
        self._txt_anonymat.setText(c.anonymat)

    def _on_ok(self):
        nom = self._txt_nom.text().strip()
        if not nom:
            QMessageBox.warning(self, "", L.get("MsgFieldRequired"))
            return
        etab_id = 0
        if self._etabs and self._combo_etab.currentIndex() >= 0:
            etab_id = self._etabs[self._combo_etab.currentIndex()].etab_id
        dob = self._date_dob.date().toString("yyyy-MM-dd")
        from app.database.repositories.candidat_repo import CandidatRepository
        c = Candidat(
            candidat_id=self._candidat_id or 0,
            nom=nom,
            nom_intermediaire=self._txt_nom_int.text().strip(),
            prenom=self._txt_prenom.text().strip(),
            date_de_naissance=dob,
            sexe=self._combo_sexe.currentText(),
            candidat_statut=self._combo_statut.currentText(),
            langue=self._combo_langue.currentText(),
            etab_id=etab_id,
            anonymat=self._txt_anonymat.text().strip()
        )
        repo = CandidatRepository()
        if self._candidat_id:
            repo.update(c)
        else:
            repo.add(c)
        self.accept()
